be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/ElectionValidator.scala

```diff
@@ -142,42 +193,184 @@ sealed class ElectionValidator(dbActorRef: => AskableActorRef) extends MessageDa
     }
   }
 
-  // checks if the votes are valid: valid question ids, valid ballot options and valid vote ids
-  private def validateVoteElection(electionId: Hash, votes: List[VoteElection], questions: List[ElectionQuestion]): Boolean = {
+  // not implemented since the back end does not receive a ResultElection message coming from the front end
+  def validateResultElection(rpcMessage: JsonRpcRequest): GraphMessage = {
+    Right(PipelineError(ErrorCodes.SERVER_ERROR.id, "NOT IMPLEMENTED: ElectionValidator cannot handle ResultElection messages yet", rpcMessage.id))
+  }
+
+  def validateEndElection(rpcMessage: JsonRpcRequest): GraphMessage = {
+    def validationError(reason: String): PipelineError = super.validationError(reason, "EndElection", rpcMessage.id)
+
+    rpcMessage.getParamsMessage match {
+      case Some(message: Message) =>
+        val (endElection, laoId, senderPK, channel) = extractData[EndElection](rpcMessage)
+
+        val electionId = channel.extractChildChannel
+
+        val firstCheck = runChecks(
+          checkTimestampStaleness(
+            rpcMessage,
+            endElection.created_at,
+            validationError(s"stale 'created_at' timestamp (${endElection.created_at})")
+          ),
+          checkId(
+            rpcMessage,
+            electionId,
+            endElection.election,
+            validationError("unexpected election id")
+          ),
+          checkId(
+            rpcMessage,
+            laoId,
+            endElection.lao,
+            validationError("unexpected lao id")
+          ),
+          checkOwner(
+            rpcMessage,
+            senderPK,
+            channel,
+            dbActorRef,
+            validationError(s"invalid sender $senderPK")
+          ),
+          checkChannelType(
+            rpcMessage,
+            ObjectType.ELECTION,
+            channel,
+            dbActorRef,
+            validationError(s"trying to send a EndElection message on a wrong type of channel $channel")
+          )
+        )
+
+        // Until runChecks() can be set to be call-by-name this two part check is required to pass the unit tests
+        lazy val secondCheck = checkVoteResults(
+          rpcMessage,
+          channel,
+          endElection.registered_votes,
+          validationError(s"Incorrect verification hash")
+        )
+
+        if (firstCheck.isLeft)
+          secondCheck
+        else
+          firstCheck
+
+      case _ => Right(validationErrorNoMessage(rpcMessage.id))
+    }
+  }
+
+  /** checks if the election has not ended
+    *
+    * @param rpcMessage
+    *   the rpc message to validate
+    * @param electionChannel
+    *   the election channel
+    * @param dbActorRef
+    *   the dbActor to ask for end election message
+    * @param error
+    *   the error to forward in case the election has ended
+    * @return
+    *   GraphMessage: passes the rpcMessages to Left if successful right with pipeline error
```

Comment:
Little typo here:
GraphMessage: passes the rpcMessages to Left if successful_, else_ right with pipeline error